### PR TITLE
Parse the contents of dynamic objects for [subobjects:false]

### LINKED ISSUE #117544: Fields not searchable with subobjects: false and dynamic: false when sent as nested objects
### Elasticsearch Version

main

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Darwin Mac 24.1.0

### Problem Description

When `subobjects: false` and `dynamic: false` are used together in an index, fields are not searchable if the documents are sent as nested objects.

### Steps to Reproduce

Prepare:
```
PUT my-index
{
  "mappings": {
    "subobjects": false,
    "dynamic": false,
    "properties": {
      "my.keyword.field": {
        "type": "keyword"
      }
    }
  }
}

POST my-index/_doc
{
  "my": {
    "keyword.field": "abc"
  }
}
```

```
GET my-index*/_search
{
  "fields": [
    "*"
  ]
}
```

Returns 
```
        "fields": {
          "my.keyword.field": [
            "abc"
          ]
        }
```

so it looks like `my.keyword.field` is indexed properly.

However,

```
GET my-index/_search
{
  "query": {
    "match": {
      "my.keyword.field": "abc"
    }
  }
}
```

does not return any hits.

* When setting `dynamic: true`, the match-search works as expected (even though the mapping looks exactly the same)
* When setting `subobjects: true`, the match-search works as expected
* When sending the doc as `{ "my.keyword.field": "abc" }`, the match-search works as expected

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by kkrik-es**
`DocumentParser` skips parsing dynamic objects when `[dynamic:false]` is set in a parent object. When `[subobjects:false]` is also set, dynamic objects should be flattened instead.

Fixes #117544

### CODE REVIEW

**Review by martijnvg — APPROVED** [review]
LGTM 👍 Thanks for fixing this Kostas!

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml [hunk 1]
```diff
@@ -1177,3 +1177,93 @@ fetch geo_point:
...
+              my.keyword.field:
+                type: keyword
+
+  - do:
+      bulk:
+        index: test
+        refresh: true
+        body:
+          - '{ "index": { } }'
+          - '{ "my": { "keyword.field": "abc" } }'
```
Comment (lkts) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml:1204]: Can we have a test that indexing a dynamic field throws in this case?

Comment (kkrik-es) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml:1204]: Done.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml [hunk 2]
```diff
@@ -1177,3 +1177,115 @@ fetch geo_point:
...
+          - '{ "my": { "random.field": "abc" } }'
+  - match: { errors: false }
+
+  - do:
+      search:
+        index: test
+        body:
+          fields: [ "*" ]
+
+  - match: { hits.hits.0.fields: { my.keyword.field: [ abc ] } }
```
Comment (lkts) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml:1223]: You'll need to sort results for this test to work with multiple shards.

File: server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java [hunk 3]
```diff
@@ -2053,6 +2053,38 @@ public void testSubobjectsFalseWithInnerDottedObject() throws Exception {
...
+            { "metrics": { "service.test": { "with.dots": "foo" }  } }"""));
+        assertNotNull(doc.rootDoc().getField("metrics.service.test.with.dots"));
+
+        doc = mapper.parse(source("""
+            { "metrics": { "service": { "test": { "with.dots": "foo" }  }  } }"""));
+        assertNotNull(doc.rootDoc().getField("metrics.service.test.with.dots"));
+
+        doc = mapper.parse(source("""
+            { "metrics": { "service": { "test.other.dots": "foo" }  } }"""));
+        assertNull(doc.rootDoc().getField("metrics.service.test.with.dots"));
```
Comment (lkts) [inline server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java:2077]:
```suggestion
        assertNull(doc.rootDoc().getField("metrics.service.test.other.dots"));
```
here and below?

**Review by lkts — APPROVED** [review]

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml (hunk 2 above)
Comment (kkrik-es) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml]: Keep forgetting that.. Done.

**Comment by kkrik-es** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.17|[](https://github.com/elastic/elasticsearch/pull/117921)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**